### PR TITLE
Monogram

### PR DESCRIPTION
**Pull request by matteocrippa**
Added support for auto monogram generation according #10 
I ported my func in objc and it picks the first and last word inside the name, so I removed also the assert for monogram and image not to be nil.

The only issue I figured out is that if you put a title in front of the name it will pick that one.

Eg.

Dr. Maria Ruiz will be DR and not MR anymore.

By the way the func check for monogram and activate the generator only it is nill.

Dunno if is better to put a check for name to be forced now not to be nil.

### CODE REVIEW

**Comment by matteocrippa** [comment]
Improved the way monogram is generated, now it makes use of a blacklist array where all words with at least a dot inside are written, this avoid the issue pointed out by @oliverschaefer in #10

**Comment by oliverschaefer** [comment]
@matteocrippa Maybe it will be clearer and easier if we add a title property to `OCKContact` that can hold the academic title (if any). Let's see what @umerkhan-apple says.

**Comment by p2** [comment]
For the blacklist, it's noteworthy that the dot-approach will also skip truncated middle names, e.g. the "A." in "James A. Corey". I guess it's not really a problem that this will become "JC" instead of "JAC".

But note that “MD” (or "PhD") is often spelled without the dots, so this approach will not skip those titles in this case.

**Comment by matteocrippa** [comment]
we can try upgrading with an extra list of blacklisted words like MD or PhD, but we have also to localise it in each country

**Comment by p2** [comment]
It could also be "MD-PhD" or "MD, PhD" or even crazier stuff like "MD, MBA, FACMI". But usually, these titles appear after a comma, so maybe just ignore everything after a comma if more than one name/word was present before the comma?

**Comment by oliverschaefer** [comment]
Wouldn't a dedicated title property be easier especially in terms of localization?

Von meinem iPhone gesendet

**Comment by matteocrippa** [comment]
Prob the best way is to split title in a new field, and have auto-monogram to generate it using only name, surname.

This will avoid each possible issue about blacklisting etc.

**Comment by matteocrippa** [comment]
@umerkhan-apple just improved the overall structure, as suggested by @oliverschaefer I pushed a new PR with a new var that will store (if exists) the academic title of the contact.
The monogram generator is still in place also.

**Comment by p2** [comment]
I think this is being over-engineered. There's no need for a title property, all you need to know is who the person on the care team is, you likely already know the person and the title can be pre/appended to the name or be omitted entirely.

As for the monogram, you most likely populate `OCKContact` from another contact element, which probably has separate fields for prefixes, suffixes, family- and given names. It's easier to generate a correct monogram from your original source than to do it inside CareKit from the full name string, so I'm not 100% convinced of the use case here either.

**Comment by oliverschaefer** [comment]
@p2 Thanks for your input. I don't think that an optional title field would instantly be an over-engineering. But as usual I'm open for discussion. What do you think @umerkhan-apple ?

**Comment by p2** [comment]
Thing is it won't work they way it's constructed right now, it assumes in two places to display `"{title} {name}"`, which is wrong for all of North America. In the Americas and Europe you can have titles before and after the name. In Asian countries it usually follows the name but can even follow the given name.

So you must at least store prefixes and suffixes in addition to a full name string, or you just keep the one “name” string. I think the latter is preferable, just assign one string like it should show up. Hell this could even be _“Dr. Bob”_ if it's your PCP.

**Comment by matteocrippa** [comment]
It's easier to set an ```enum``` like:

```
enum TitlePosition {
case .None,
case .Prefix,
case .Suffix
}
```

 this will allow developer to manage this in the better way according localisation

**Comment by oliverschaefer** [comment]
@p2 You are right, we cannot append the title to the name (or the other way around). The easiest thing would be to display the title separately (e.g. as a detailTextLabel in a UITableViewCell). That would fulfill the demand of entering a title (which some of my physician-colleagues are quite strict about) and also keep things easy in terms of localization and monogram generation (which this topic originally was all about ;-) )

**Comment by p2** [comment]
I still maintain having only a `name` property that includes all titles is by far the best solution for CareKit.

**Comment by oliverschaefer** [comment]
@p2 Probably yes, but killing the monogram auto-generation feature.

**Comment by matteocrippa** [comment]
@p2 @oliverschaefer at this very moment it acts like iOS ABAddressBook:
- you can set a Name and Surname
- you can set a Title

But it generates the monogram without the title and append it before the name

img_1858

img_1857

**Comment by p2** [comment]
@oliverschaefer Ignoring everything ending with a period and everything after the first comma – probably only if there are at least two words before the comma – could still result in good monogram auto-creation. For edge cases the app developer can just create the monogram herself.

**Comment by umerkhan-apple** [comment]
@matteocrippa If the auto-generation will have edge cases that cannot be handled by the implementation, then it might be best to just ask the developer for the monogram. 

Currently, the monogram is non-null, we could potentially make it null_resettable. If the developer does not provide one, we can auto-generate one for them.

@p2 @oliverschaefer Let me know how you guys feel about this approach.

**Comment by p2** [comment]
I feel making it `null_resettable` and auto-generating one if it's null with @matteocrippa 's approach would be reasonable. There will be edge cases, then it's up to the developer to work around those to improve her/his app.

**Comment by oliverschaefer** [comment]
Agree

Von meinem iPhone gesendet

**Comment by matteocrippa** [comment]
I was looking at [CNContact](https://developer.apple.com/library/ios/documentation/Contacts/Reference/CNContact_Class/index.html#//apple_ref/occ/instp/CNContact/namePrefix) that will be for sure in some cases the real source of the data the user is going to provide.

You can see there that there are two vars, one for **nameSuffix** and one for **namePrefix** why not approach it in this way or (probably) even better use ```CNContact``` instance or reference and add only the missing fields to the data model (e.g. relation, tintColor)?

In this way monogram will be generated easily looking at **.givenName** and **.familyName** and the user will also be able to easily store the information inside his/her local address book too.

**Comment by matteocrippa** [comment]
b8b8b0a provides monogram as ```null_resettable```

**Comment by p2** [comment]
@matteocrippa That is what I meant, structured names usually have prefixes and suffixes. I don't think it makes sense to have yet another structured way of maintaining contact names in CareKit. But, often, the source of the name is structured, either `CNContact` or some API return. From there it's easy to create a monogram because you only look at given, middle and family names. And if you don't create a monogram, the suggestions posited above are good enough IMHO.

**Comment by umerkhan-apple** [comment]
@matteocrippa I think we should take out the `title` property from `OCKContact`. A developer can use a formatter to create the full name string and provide it to the `OCKContact` object. Though we should keep the `null_resettable` on the monogram. We also won't need a blacklist, because if the developer would like a monogram other than the auto-generated one, they can specify one.

**Comment by matteocrippa** [comment]
@umerkhan-apple pushed new commits, now it should work as you suggested

File: CareKit/Connect/OCKContact.m [hunk 1]
```diff
@@ -134,4 +133,37 @@ - (instancetype)copyWithZone:(NSZone *)zone {
     return contact;
 }
 
+#pragma mark - Monogram
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKContact.m]: add new line

File: Sample/OCKSample/SampleData.swift [hunk 2]
```diff
@@ -57,67 +57,67 @@ class SampleData: NSObject {
             phoneNumber: CNPhoneNumber(stringValue: "888-555-5512"),
             messageNumber: CNPhoneNumber(stringValue: "888-555-5512"),
             emailAddress: "mruiz2@mac.com",
-            monogram: "MR",
```
Comment (umerkhan-apple) [inline Sample/OCKSample/SampleData.swift]: Lets keep this one to test both use-cases.

File: CareKit/Connect/OCKConnectTableViewCell.m [hunk 3]
```diff
@@ -106,6 +106,8 @@ - (void)updateView {
     }
     
     _nameLabel.text = self.contact.name;
+    
+
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKConnectTableViewCell.m]: Remove extra new lines.

**Comment by umerkhan-apple** [comment]
@matteocrippa Minor nits and merge conflicts. Apart from that, looks good to go! 👍

File: CareKit/Connect/OCKContact.m [hunk 4]
```diff
@@ -134,4 +133,37 @@ - (instancetype)copyWithZone:(NSZone *)zone {
...
+- (void)setMonogram:(NSString *)monogram {
+    if (!monogram) {
+        monogram = [self generateMonogram:_name];
```
Comment (p2) [inline CareKit/Connect/OCKContact.m]: Could this be implemented on the getter, not the setter? Then the `self.monogram` "hack" in the initializer is not needed.

**Comment by matteocrippa** [comment]
updated my fork with master and merged in my branch

File: CareKit/Connect/OCKConnectTableViewCell.m [hunk 5]
```diff
@@ -101,12 +101,19 @@ - (void)updateView {
...
     } else {
         _imageView.image = nil;
-        _monogramLabel.text = self.contact.monogram;
+        _imageView.backgroundColor = OCKSystemGrayColor();
+        
+        NSString *monogram = self.contact.monogram;
+        NSRange stringRange = {0, MIN([monogram length], 2)};
+        stringRange = [monogram rangeOfComposedCharacterSequencesForRange:stringRange];
+        NSString *clippedMonogram = [monogram substringWithRange:stringRange];
+        _monogramLabel.text = clippedMonogram;
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKConnectTableViewCell.m]: PR https://github.com/carekit-apple/CareKit/pull/10 takes care of the clipping as part of `OCKContact`.

File: CareKit/Connect/OCKConnectTableViewCell.m [hunk 6]
```diff
@@ -101,12 +101,19 @@ - (void)updateView {
...
     }
     
     [self updateImageViewBackgroundColor];
     
     _nameLabel.text = self.contact.name;
+
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKConnectTableViewCell.m]: Remove extra line.

File: CareKit/Connect/OCKConnectTableViewCell.m (hunk 5 above)
Comment (umerkhan-apple) [inline CareKit/Connect/OCKConnectTableViewCell.m]: PR https://github.com/carekit-apple/CareKit/pull/10 takes care of the monogram clipping as part of `OCKContact`.

File: CareKit/Connect/OCKContact.m [hunk 7]
```diff
@@ -49,8 +49,7 @@ - (instancetype)initWithContactType:(OCKContactType)type
                        emailAddress:(NSString *)emailAddress
                            monogram:(NSString *)monogram
                               image:(UIImage *)image {
-    NSAssert((monogram || image), @"An OCKContact must have either a monogram or an image.");
-    
+
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKContact.m]: Remove extra line.

File: CareKit/Connect/OCKContact.m [hunk 8]
```diff
@@ -143,4 +133,44 @@ - (instancetype)copyWithZone:(NSZone *)zone {
...
+- (void)setMonogram:(NSString *)monogram {
+    if (!monogram) {
+        monogram = [self generateMonogram:_name];
+    }
+    _monogram = [monogram copy];
+}
+
+- (NSString *)generateMonogram:(NSString *)name {
+
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKContact.m]: Remove extra line.

**Comment by matteocrippa** [comment]
sorry, now it should be fixed

**Comment by matteocrippa** [comment]
hey @umerkhan-apple this PR is still pending, are you considering to merge or do you prefer to change the approach for monogram?

File: Sample/OCKSample/SampleData.swift [hunk 9]
```diff
@@ -59,65 +59,65 @@ class SampleData: NSObject {
...
             messageNumber: CNPhoneNumber(stringValue: "888-555-5512"),
             emailAddress: "billjames2@mac.com",
-            monogram: "BJ",
+            monogram: nil,
             image: nil),
-        
+
         OCKContact(contactType: .Personal,
-            name: "Tom Clark",
+            name: "Tom Van de Clark",
```
Comment (umerkhan-apple) [inline Sample/OCKSample/SampleData.swift]: Can we remove Van de?

**Comment by umerkhan-apple** [comment]
@matteocrippa Apart from one minor nit. This looks good!
@YuanZhu-apple Can you take a look as well?

File: Sample/OCKSample/SampleData.swift (hunk 9 above)
Comment (matteocrippa) [inline Sample/OCKSample/SampleData.swift]: I put that only to check the issue w/ multiple words inside an user name, but we can for sure remove it.

Comment (umerkhan-apple) [inline Sample/OCKSample/SampleData.swift]: That makes sense. Let's go ahead and put move the check into the test app/

**Comment by umerkhan-apple** [comment]
@YuanZhu-apple Friendly ping! :)

**Pull request merged**